helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -1299,6 +1308,10 @@ protected void checkLiveInstancesObservation(List<LiveInstance> liveInstances,
   }
 
   public void shutdown() throws InterruptedException {
+    if (_periodicRebalancerFutureTasks != null) {
+      _periodicRebalancerFutureTasks.cancel(false);
+    }
+    _periodicalRebalanceExecutor.shutdown();
```

Comment:
Does this shutdown kill the thread immediately? I think not. 

Shall we provide an shutDownNow which kill the thread immediately? That would be useful for tests.

Reply:
Yea it does not kill the thread immediately. There is a ScheduledExecutorService.shutDownNow() that interrupt threads and we could use that in test I guess.

Reply:
`   _asyncTasksThreadPool.shutdownNow();
    try {
      _asyncTasksThreadPool.awaitTermination(EVENT_THREAD_JOIN_TIMEOUT, TimeUnit.MILLISECONDS);
    } catch (InterruptedException ex) {
      logger.warn("Timeout when terminating async tasks. Some async tasks are still executing.");
    }
`
see the above. Can we do something similar like the above?

Reply:
shutdownNow cancels already scheduled TimerTask and interrupt running TimerTask. For this  _periodicalRebalanceExecutor, scheduled TimerTask is canceled in stopPeriodRebalance() and if there is running one, it is interrupted in shutdownPipeline();

Reply:
Please correct me if I am wrong. In my understanding, shutdownPipeline() guarantees no event processor thread is alive anymore.

Reply:
Yes shutdownPipeline() does shut down the event process thread, but when you call shutdownNow, it's the thread inside the executor is shut down. I think they are different threads.

Reply:
TFTR. Updated.